idea/idea-frontend-api/src/org/jetbrains/kotlin/idea/frontend/api/KtSmartCastNullability.kt

```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.idea.frontend.api
+
+/**
+ * The nullability of an expression, based on smart cast types derived from data-flow analysis facts. Example:
+ * ```
+ *   fun foo(s: String?) {
+ *     s  // KtSmartCastNullability of `s` here is UNKNOWN
+ *     if (s != null) {
+ *       s  // KtSmartCastNullability.NOT_NULL
+ *     } else {
+ *       s  // KtSmartCastNullability.NULL
+ *     }
+ *     s!!
+ *     // KtSmartCastNullability of `s` from this point is NOT_NULL
+ *   }
+ * ```
+ * Note that the nullability returned here is from "stable" smart cast types. The
+ * [spec](https://kotlinlang.org/spec/type-inference.html#smart-cast-sink-stability) provides an explanation on smart cast stability.
+ */
+enum class KtSmartCastNullability(val canBeNull: Boolean, val canBeNonNull: Boolean) {
```

Comment:
Or, maybe we can introduce `isDefinitelyNull()` (as well as `isDefinitelyNotNull()`) utils to avoid direct enum entry comparions.

Reply:
Ok, I'll go with that. My concern with any of these solutions is that we're introducing a concept that isn't documented, whatever we call it. Whether it's "stable", "smart cast", or "definite" nullability. I went with "stable" at first, then "smart cast" since there was a little bit of documentation about it. But "definite" nullability seems more straightforward to explain and can be used for regular non-nullable types.

Reply:
To be clear, I meant I will go with just introducing `isDefinitelyNull()` and `isDefinitelyNotNull()`

Reply:
Thanks! `isDefinitelyNull` and `isDefinitelyNotNull` sound good to me too!